pkg/core/resources/apis/meshservice/match.go

```diff
@@ -0,0 +1,114 @@
+		case ms.Spec.Selector.DataplaneTags != nil:
+			result[ms] = matchByTags(ms, dppsByNameByTag)
+		default:
+			result[ms] = nil
+		}
+	}
+
+	return result
+}
+
+func indexDpsForMatching(
+	dpps []*core_mesh.DataplaneResource,
+	matchOnlyHealthy bool,
+) (
+	map[string]map[string]*core_mesh.DataplaneResource,
+	map[string]*core_mesh.DataplaneResource,
+) {
+	// Map of tag pair in a context of a mesh (ex. app:redis from default mesh) to dpp name to dpp.
+	dppsByNameByTag := map[string]map[string]*core_mesh.DataplaneResource{}
+	dppsByName := map[string]*core_mesh.DataplaneResource{}
+
+	for _, dpp := range dpps {
+		inbounds := dpp.Spec.GetNetworking().GetInbound()
+		if matchOnlyHealthy {
+			inbounds = dpp.Spec.GetNetworking().GetHealthyInbounds()
+		}
+
+		for _, inbound := range inbounds {
+			for tagName, tagValue := range inbound.GetTags() {
+				tag := fmt.Sprintf("%s;%s;%s", dpp.Meta.GetMesh(), tagName, tagValue)
+				dataplanes, ok := dppsByNameByTag[tag]
+				if !ok {
+					dataplanes = map[string]*core_mesh.DataplaneResource{}
+					dppsByNameByTag[tag] = dataplanes
+				}
+				dataplanes[dpp.Meta.GetName()] = dpp
+			}
+		}
+		if len(inbounds) > 0 {
+			dppsByName[fmt.Sprintf("%s;%s", dpp.Meta.GetMesh(), dpp.Meta.GetName())] = dpp
+		}
+	}
+	return dppsByNameByTag, dppsByName
+}
+
+func matchByRef(
+	ms *meshservice_api.MeshServiceResource,
+	dppsByName map[string]*core_mesh.DataplaneResource,
+) []*core_mesh.DataplaneResource {
+	key := fmt.Sprintf("%s;%s", ms.Meta.GetMesh(), ms.Spec.Selector.DataplaneRef.Name)
+	if dpp, ok := dppsByName[key]; ok {
+		return []*core_mesh.DataplaneResource{dpp}
+	}
+	return nil
+}
+
+func matchByTags(
+	ms *meshservice_api.MeshServiceResource,
+	dppsByNameByTag map[string]map[string]*core_mesh.DataplaneResource,
+) []*core_mesh.DataplaneResource {
+	// Find the shortest map of dpps by name that matches the tags
+	// For example, if we have MeshService with selector of `app: redis` and `k8s.kuma.io/namespace: kuma-demo`
+	// and DPPs:
+	// * `app: redis`, `k8s.kuma.io/namespace: kuma-demo`
+	// * `app: demo-app`, `k8s.kuma.io/namespace: kuma-demo`
+	// It's better to grab the shortest list of DPPs (app:redis) to reduce number of operations (.Matches) for the next step.
+	var shortestDppMap map[string]*core_mesh.DataplaneResource
+	for tagName, tagValue := range ms.Spec.Selector.DataplaneTags {
+		tagsKey := fmt.Sprintf("%s;%s;%s", ms.GetMeta().GetMesh(), tagName, tagValue)
+		if dppsByName, ok := dppsByNameByTag[tagsKey]; ok {
+			if shortestDppMap == nil || len(dppsByName) < len(shortestDppMap) {
+				shortestDppMap = dppsByName
+			}
+		} else {
+			// No proxies will match this pair of tags, no point in going further.
+			shortestDppMap = nil
+			break
+		}
+	}
+
+	// Go over the shortest list of data plane proxies and pick only proxies that matches all tags.
+	var dpps []*core_mesh.DataplaneResource
+	for _, dppName := range maps.SortedKeys(shortestDppMap) {
```

Comment:
Question: Do we need a map to store DPPs?

Current approach:

1. Iterate over sorted keys.
2. Retrieve the DPP for each key.
3. Check for matches and append if they match.
4. Repeat the process.

Could we achieve the same result by only iterating over the list of DPPs? Is there any specific reason to keep them in a map?

Reply:
Great thanks! Let's note it so we won't change it in the future.

Reply:
added a comment. Unit test already covers it